src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -845,8 +846,8 @@ namespace shogun
 		 * represented
 		 * as \f$\sum_i a_i b_i\f$
 		 */
-		template <typename T>
-		T dot(const SGVector<T>& a, const SGVector<T>& b)
+		template <typename T, typename U, typename TU = typename promote<T, U>::type>
+		TU dot(const SGVector<T>& a, const SGVector<U>& b)
```

Comment:
a more general purpose dot product interface. but the backend only implements all combinations with T=U and with T=float64_t, U != T